### ISSUE TITLE
Tracking issue: non-deterministic java apps

### ISSUE THREAD

**Issue by TomaSajt** [NixOS/nixpkgs#278518]
There are several ways to package a Java app inside Nixpkgs, almost all of them requiring the use of calling the `java` command with the produced `.jar` files.

One could just download the pre-built `.jar`-s and not worry about this, but I think it's important to build stuff from source.

Here are some ways packages build java apps:
- using `javac` by itself
- using `ant/apacheAnt`
- using `maven` (with `maven.buildMavenPackage`)
- using `gradle` (not many pacakges do this inside Nixpkgs currently, as there's no generic builder for it yet)

### Why is it non-deterministic?

When creating a `.jar` file Java will place a `META-INF/MANIFEST.MF` file inside the archive which will have the current date as its creation date, so it's non-deterministic.

[How to check for reproducibility?](https://github.com/NixOS/nixpkgs/blob/master/.github/ISSUE_TEMPLATE/unreproducible_package.md)

There also might be a problem with the fact, that changing a `.properties` file in java usually with `ant`, the current timestamp will be put inside the file as a comment.

### Solution

There has been an attempt at creating a setup-hook called `canonicalize-jars-hook` which aims to solve this problem by unwrapping and rewrapping jars with the creation dates set to a fixed time.

Maybe this hook should be added to the `java` packaging docs

I discovered some issues with it though, which I address in this PR: https://github.com/NixOS/nixpkgs/pull/278322

I went ahead refactored some packages to use this hook so that they can be deterministic.
I will be opening some PRs with the changes
Here's my progress:

- [x] abcl (PR: ---)
- [x] calcoo (PR: ---)
- [x] dayon (PR: ---)
- [x] hydra-ant-logger (PR: ---)
- [x] ili2c (PR: ---)
  - .properties file use patched
- [x] jasmin (PR: ---)
- [x] javacc (PR: ---)
- [x] javaCup (PR: ---)
- [x] jdepend (PR: ---)
- [x] jna (PR: ---)
- [x] jpsxdec (PR: ---)
- [x] jugglinglab (PR: ---)
  - version update
  - moved to maven from ant
- [x] martyr (PR: ---)
- [x] pattypan (PR: ---)
- [x] trimmomatic (PR: ---)
- [x] tvbrowser (PR: ---)

### has broken linebreaks, should use fixed hook from https://github.com/NixOS/nixpkgs/pull/278322
- [x] dataexplorer (PR: ---)
  - version update
- [x] domination (PR: ---)
- [x] fod (PR: ---)
- [x] freetts (PR: ---)
- [x] projectlibre (PR: ---)
  - also has overlapped files

---
- [ ] axis2
  - not attempted
  - has .war files
- [ ] brmodelo
  - not attempted
- [ ] i2p
  - not deterministic (unstable xml order inside .war files)
- [ ] freemind
  - not deterministic (.properties file being edited)
- [ ] gcs
  - not attempted
  - moved to Golang in 5.0.0, not important for this PR
- [ ] jameica
  - not attempted
- [ ] jedit
  - not attempted
- [ ] jffi
  - WIP
- [ ] jogl
  - not attempted
- [ ] mkgmap
  - not attempted
- [ ] processing
  - not attempted
- [ ] rabbitmq-java-client
  - not attempted
  - uses python2
- [ ] sweethome3d
  - not attempted
- There might be other packages that build jars which I didn't list here

**Comment by fgaz** [comment]
> Maybe this hook should be added to the java packaging docs

yes please!

**Comment by de11n** [comment]
Excellent work.

**Comment by TomaSajt** [comment]
I was looking around on nixpkgs and it looks like `ant` actually does have a way to set the modification-time for the created jars. Here's the only example that's inside nixpkgs:
https://github.com/NixOS/nixpkgs/blob/6f5673f731614db0f22d9af504ac844a1112a17c/pkgs/development/java-modules/jogl/default.nix#L50-L52
It's editing the adding a fixed `modificationtime` attribute to the `jar` task.
Docs of the `jar` ant task: https://ant.apache.org/manual/Tasks/jar.html

After some more looking, it turns out that even the `jar` program itself has the `--date` argument.
```
jar --date="1980-01-01T00:00:02Z" --create Program.class > test.jar
```
This will create a jar file with all files inside having the set timestamp. Though this is not in the older versions of java. The first version to have this inside nixpkgs is `jdk17`. Not sure exactly which version added it because it is heavily underdocumented.

After some more searching, it looks like `gradle` also has something like this built in:
```
tasks.withType(AbstractArchiveTask) {
    preserveFileTimestamps = false
    reproducibleFileOrder = true
}
```

---
So, in conclusion, it looks like every major way of packaging has a built-in solution for fixed timestamps. I am a bit sad that I did not look into this more thoroughly earlier.
Still, I believe that having a hook, which works well can still be useful, as it will allow us to avoid having to patch every java app.

**Comment by TomaSajt** [comment]
I decided that using the built-in method is a better solution, so I opened https://github.com/NixOS/nixpkgs/pull/294516, which moves away from `canonicalize-jars-hook`

**Comment by de11n** [comment]
It might also be helpful to have a tool to detect common pitfalls in making Java packages deterministic. For example, we could find all jar files in the outputs and inspect them for timestamps that suggest non-determinism. We could then cause that build to fail unless some sort of "allowNondetermism = true" flag is set. This could be a setuphook but at some point we may want a `buildJavaPackage` builder that just applies common-sense things like this and can be easily documented.

**Comment by TomaSajt** [comment]
A new thing I discovered was the existance of `.jmod` files. AFAICT they are present inside every `jdk`. `strip-nondeterminism` supports patching these files, however I am not sure if it's the best idea to have all `jdk`s as a dependency of `strip-nondeterminism`
